datalad_next/download_file.py

```diff
@@ -0,0 +1,254 @@
+    #spec=spec_item_constraint | EnsureListOf(spec_item_constraint)# \
+    spec_constraint = AltConstraints(
+        spec_item_constraint,
+        EnsureListOf(spec_item_constraint),
+        EnsureGeneratorFromFileLike(spec_item_constraint),
+    )
+
+    force_choices = EnsureChoice('overwrite-existing')
+
+    # Interface.validate_args() will inspect this dict for the presence of a
+    # validator for particular parameters
+    _validators_ = dict(
+        spec=spec_constraint,
+        # if given, it must also exist as a source for configuration items
+        # and/or credentials
+        dataset=EnsureDataset(installed=True),
+        force=force_choices | EnsureListOf(force_choices),
+    )
+
+    # this is largely here for documentation and CLI parser building
+    _params_ = dict(
+        spec=Parameter(
+            args=('spec',),
+            doc="""""",
+            nargs='+'),
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc=""""""),
+        force=Parameter(
+            args=("--force",),
+            action='append',
+            doc=""""""),
+    )
+
+    _examples_ = [
+    ]
+
+    @classmethod
+    def validate_args(cls: Interface, kwargs: Dict, at_default: set) -> Dict:
+        validated = {}
+        for argname, arg in kwargs.items():
+            if argname in at_default:
+                # do not validate any parameter where the value matches the
+                # default declared in the signature. Often these are just
+                # 'do-nothing' settings or have special meaning that need
+                # not be communicated to a user. Not validating them has
+                # two consequences:
+                # - the condition can simply be referred to as "default
+                #   behavior" regardless of complexity
+                # - a command implementation must always be able to handle
+                #   its own defaults directly, and cannot delegate a
+                #   default value handling to a constraint
+                #
+                # we must nevertheless pass any such default value through
+                # to make/keep them accessible to the general result handling
+                # code
+                validated[argname] = arg
+                continue
+            validator = cls._validators_.get(argname, lambda x: x)
+            # TODO option to validate all args despite failure
+            try:
+                validated[argname] = validator(arg)
+            except Exception as e:
+                raise ValueError(
+                    f'Validation of parameter {argname!r} failed') from e
+        return validated
+
+    @staticmethod
+    @datasetmethod(name="download_file")
+    @eval_results
+    def __call__(spec, *, dataset=None, force=None):
+        # which config to inspect for credentials etc
+        cfg = dataset.ds if dataset else datalad.cfg
+
+        http_handler = HttpOperations(cfg)
```

Comment:
Maybe have the handler definitions outside of the command, so they could be imported elsewhere?

Generally, I'm not yet convinced that special remotes should actually call the command but its helpers (like those handlers), because the command comes with an overhead that would apply per key from within a special remote.

Reply:
This is https://github.com/datalad/datalad-next/issues/129

> Generally, I'm not yet convinced that special remotes should actually call the command

I hope there is nothing in the code that is *trying* to convince you ;-)

Reply:
> I hope there is nothing in the code that is trying to convince you ;-)

Nope, just recalled a suggestion by you elsewhere.
Seems we are on the same page then.